### PR TITLE
 XML compression

### PR DESCRIPTION
**Pull request by satyaakam**
https://dimagi-dev.atlassian.net/browse/IIO-742
##### SUMMARY

Changes in BlobDB API were made as per the suggestions on https://github.com/dimagi/commcare-hq/issues/26813
##### FEATURE FLAG

ICDS-CAS

##### PRODUCT DESCRIPTION

None

### CODE REVIEW

File: corehq/blobs/s3db.py
Comment (snopoke) [inline corehq/blobs/s3db.py:76]: can we avoid doing the compression twice?

```suggestion
            if meta.compressed:
                buf = io.BytesIO()
                with GzipFile(fileobj=buf, mode='wb') as f:
                    f.write(data)
                content = buf
            meta.content_length = get_file_size(content)
```
I realize that this will load all the data in memory but that's probably OK for our current use cases (at least it would simplify things initially and we could optimize if we need to compress larger objects). If we did this we could get rid of `GzipCompressReadStream` completely

File: corehq/blobs/interface.py
Comment (snopoke) [inline corehq/blobs/interface.py:57]: should update the docstring with details of how to call this

File: corehq/blobs/interface.py
Comment (snopoke) [inline corehq/blobs/interface.py:82]: I think the version here is more concise: https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727

File: corehq/blobs/util.py
Comment (snopoke) [inline corehq/blobs/util.py:76]: This could do with some tests. Also wondering if it wouldn't be cleaner to implement this as a file wrapper using the zlip encoder directly. Something similar to this: http://effbot.org/librarybook/zlib/zlib-example-4.py

File: corehq/blobs/s3db.py
Comment (snopoke) [inline corehq/blobs/s3db.py:82]: need to add the args here

File: corehq/form_processor/models.py
Comment (snopoke) [inline corehq/form_processor/models.py:1077]: I don't think this is necessary as `CaseAttachments` will always be `CODES.form_attachment`. Also we don't really support case attachments (it's on the kill list)

**Review by snopoke — COMMENTED** [review]
I did a quick pass and left some comments. Overall looks like a good start.

I don't think you meant to commit these:
image

File: corehq/blobs/s3db.py
Comment (millerdev) [reply corehq/blobs/s3db.py:76]: @snopoke I would prefer to maintain support for arbitrarily large files. The blob db has been designed that way from the beginning, and I think it's an important part of the mental model for building things on top of it.

My recommendation for accomplishing that would be to have `GzipCompressReadStream` keep track of how many bytes have been read
```py
            if meta.compressed:
                content = GzipCompressReadStream(content)
            with self.report_timing('put', meta.key):
                s3_bucket.upload_fileobj(content, meta.key)
            meta.content_length = get_file_size(content)
            self.metadb.put(meta)
        return meta

...

def get_file_size(fileobj):
    if isinstance(fileobj, GzipCompressReadStream):
        return fileobj.content_length
```
`<GzipCompressReadStream>.content_length` implementation TBD. The attribute should not be set initially (it is an error to read the content length before all bytes have been read). It should be set to an integer value once all bytes have been read from the input.

Caveat: I can't remember if it's important to save the metadata _before_ writing blob content to S3. If it is then we have a big problem since we could never optimize this for arbitrary file size.

However, there is another question I'm not sure we've considered: is `meta.content_length` the number of bytes stored in the backend, or is it the number of bytes that would be returned by reading the file-like object returned by `blobdb.get(...)`? I can see reasons why we would want both, and it seems possibly important to have it be the latter since the blob db handles compression transparently. Thoughts?

File: corehq/blobs/util.py
Comment (millerdev) [inline corehq/blobs/util.py:58]:
```suggestion
        while size > 0 and self.__buf:
```

File: corehq/blobs/util.py
Comment (millerdev) [inline corehq/blobs/util.py:46]: Private members should be prefixed with one underscore, not two. Name mangling (invoked by dunder-prefix) is nasty and should be avoided if possible. The same applies to `GzipCompressReadStream` below.

File: corehq/blobs/util.py
Comment (millerdev) [inline corehq/blobs/util.py:43]: nit: Python 3 ftw!
```suggestion
class Buffer:
```

File: corehq/blobs/util.py
Comment (millerdev) [inline corehq/blobs/util.py:73]:
```suggestion
        self.__buf = None
```

File: corehq/blobs/util.py
Comment (millerdev) [inline corehq/blobs/util.py:85]:
```suggestion
        while size < 0 or self.__buf < size:
```

Comment (millerdev) [reply corehq/blobs/util.py:86]: Ignore this. I don't know what I was thinking here.

File: corehq/blobs/migrations/0011_blobmeta_compressed.py
Comment (snopoke) [inline corehq/blobs/migrations/0011_blobmeta_compressed.py:8]: must be annotated with `@partitioned` (`corehq.sql_db.migrations.partitioned`)

File: corehq/blobs/interface.py
Comment (satyaakam) [reply corehq/blobs/interface.py:89]: Following issues with that version:
- If function is called with `type_code == "form_xml"` and one of `key` or `error`, they will get following error:
```
'key' must be specified with 'type_code'
```
whereas what they really should get is
```
form XML must be loaded with 'meta' argument
```
Current version, although longer, handles all possible cases with the correct error message

File: corehq/blobs/util.py
Comment (satyaakam) [reply corehq/blobs/util.py]: test cases written in test_util.py

File: corehq/blobs/interface.py
Comment (millerdev) [reply corehq/blobs/interface.py:89]: I didn't follow this. Are you saying this function call
```py
validate_get_args("some-key", CODES.form_xml, None)
```
will raise `'key' must be specified with 'type_code'`?

That's not what I see. Tested locally this is what I get:
```
Traceback (most recent call last):
  File "<string>", in validate_get_args
ValueError: form XML must be loaded with 'meta' argument

```

File: corehq/blobs/s3db.py
Comment (millerdev) [inline corehq/blobs/s3db.py:199]: Is this used?

File: corehq/blobs/util.py
Comment (millerdev) [inline corehq/blobs/util.py:58]: This should use a custom exception subclass rather than `Exception`. As is, it is impossible to catch only this error.

File: corehq/blobs/tests/test_util.py
Comment (millerdev) [inline corehq/blobs/tests/test_util.py:50]: Please add another test: write several bytes, then do a partial read (say write 10 bytes, then read 5), verify that exception is raised on access `compress_stream.content_length` before all 10 bytes have been read.

File: corehq/blobs/interface.py
Comment (satyaakam) [reply corehq/blobs/interface.py:89]: No, I am referring to the version at https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727 since @snopoke suggested using it.

Using that version will raise `'key' must be specified with 'type_code'` with following function call:

validate_get_args("some-key", CODES.form_xml, None)

Comment (millerdev) [reply corehq/blobs/interface.py:89]: Might want to check that again. I ran the version from https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727 and got `ValueError: form XML must be loaded with 'meta' argument`

```py
>>> class CODES:
...     # stub
...     form_xml = 2
...
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
...
>>> validate_get_args("some-key", CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 8, in validate_get_args
ValueError: form XML must be loaded with 'meta' argument
```

Comment (satyaakam) [reply corehq/blobs/interface.py:89]: Ok, I checked again. But consider below

>>> class CODES:
...     form_xml = 2
... 
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
... 
>>> validate_get_args(None, CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in validate_get_args
ValueError: 'key' must be specified with 'type_code'

Comment (satyaakam) [reply corehq/blobs/interface.py:89]: Ok, I checked again. But consider below

```
>>> class CODES:
...     form_xml = 2
... 
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
... 
>>> validate_get_args(None, CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in validate_get_args
ValueError: 'key' must be specified with 'type_code'
```

Comment (millerdev) [reply corehq/blobs/interface.py:89]: Looks right to me (you sent a null key with a type code). Do you see a problem?

Comment (satyaakam) [reply corehq/blobs/interface.py:89]: "We sent `type_code` as `form_xml`. `key` is not expected for this type_code and only `meta` should be provided. Yet the error asks me to provide `key`."

Comment (millerdev) [reply corehq/blobs/interface.py:89]: This should fix it
```py
def validate_get_args(key, type_code, meta):
    if key is not None or type_code is not None:
        if meta is not None:
            raise ValueError("'key' and 'meta' are mutually exclusive")
        if type_code == CODES.form_xml:
            raise ValueError("form XML must be loaded with 'meta' argument")
        if key is None or type_code is None:
            raise ValueError("'key' must be specified with 'type_code'")
        return key
    if meta is None:
        raise ValueError("'key' and 'type_code' or 'meta' is required")
    return meta.key
```

File: corehq/blobs/s3db.py
Comment (snopoke) [reply corehq/blobs/s3db.py]: > However, there is another question I'm not sure we've considered: is meta.content_length the number of bytes stored in the backend (compressed bytes), or is it the number of bytes that would be returned by reading the file-like object returned by blobdb.get(...) (uncompressed bytes)? I can see reasons why we would want both, and it seems possibly important and more intuitive to have it be the latter since the blob db handles compression transparently. Thoughts?

I was wondering about this too. We don't really use this field very much but in the few places where it is used we need the uncompressed size (except where we post metrics to Datadog where we'd want the compressed size). One place were we wouldn't be able to get the uncompressed size is where we copy an existing blob: https://github.com/dimagi/commcare-hq/blob/8f63335b220751b0bdaac9e13abb818e0ab02413/corehq/blobs/s3db.py#L70. This appears to have been added specifically for case attachments here: https://github.com/dimagi/commcare-hq/pull/13203. I couldn't find any usages of it but there must be some since we're recording metrics for it on [production](https://app.datadoghq.com/dashboard/753-35q-fwt/blob-db-dashboard?from_ts=1584428385517&fullscreen_section=overview&fullscreen_widget=3073215485582670&live=true&tile_size=m&to_ts=1587020385517&tpl_var_env=production&utc_override=true&fullscreen_start_ts=1584446132694&fullscreen_end_ts=1587038132694&fullscreen_paused=false). Either way that does seem like a useful feature to keep. One approach would be to force the user to pass in the `meta` to `db.put` if they are passing in an existing blob or else pass the content length info into `BlobStream`.

I also noticed that if we return a `GzipFile` from `db.get` then the type check `isinstance(content, BlobStream)` will fail and we'll end up to a read then write.

Comment (snopoke) [reply corehq/blobs/s3db.py]: @millerdev I've made some updates here so that `meta.content_length` is the uncompressed size but there are 2 remaining challenges:

1. `db.size(key) != meta.content_length`

This is used in a few places to to add content length headers to responses. I think this could be solved by applying the same arg rules to `size` as we have to `get`. We could also eliminate all calls to `size` if we return the content length along with the blob from `get` since they are all in places where the actual blob is fetched as well.

2. The metrics are reporting the uncompressed size which I think is incorrect.

Not clear how to solve this since many places would not be able to get the compressed size easily e.g. bulk delete

File: corehq/blobs/models.py
Comment (millerdev) [inline corehq/blobs/models.py:107]: This can be simplified
```suggestion
        from . import get_blob_db
        db = db or get_blob_db()
        return db.get(meta=self)
```

**Review by millerdev — APPROVED** [review]
Looking good. One small recommendation.

File: corehq/blobs/s3db.py
Comment (millerdev) [reply corehq/blobs/s3db.py]: Have you considered adding a `compressed_size` attribute to `BlobMeta`? It would be null if `<BlobMeta>.compressed` is false. We could even improve storage efficiency a bit by removing `compressed` entirely and make the rule be that if `compressed_size` is null then the blob is not compressed, otherwise it is compressed. Thoughts?

Storing both `compressed_size` in addition to `content_length` is beginning to seem more straightforward than other options.

File: corehq/blobs/metadata.py
Comment (snopoke) [inline corehq/blobs/metadata.py:36]: I don't really like this. It does got overwritten during later but still feels off

File: corehq/blobs/metadata.py
Comment (snopoke) [reply corehq/blobs/metadata.py:20]: Used here: https://github.com/dimagi/commcare-hq/blob/e0d7c553ea27a08146acbb9b4b4a9a39753b7b16/corehq/blobs/mixin.py#L193

File: corehq/blobs/s3db.py
Comment (snopoke) [reply corehq/blobs/s3db.py:88]: I don't thinks so:
```python

def get_content_size(fileobj, chunks_sent):
    if isinstance(fileobj, GzipCompressReadStream):
        return fileobj.content_length, sum(chunks_sent)

    return sum(chunks_sent), None
```
So when it's compressed then it returns the uncompressed size and the sum of the chunks sent which should be the compressed size.

Comment (snopoke) [reply corehq/blobs/s3db.py:88]: Note that the old implementation doesn't work in most cases since the file get's closed by `s3_bucket.upload_fileobj`.

Comment (millerdev) [reply corehq/blobs/s3db.py:88]: `fileobj.content_length` (where `fileobj` is `GzipCompressReadStream`) returns compressed size

https://github.com/dimagi/commcare-hq/blob/820f81d1249157f34b592bebdb5d6b8868ae461f/corehq/blobs/util.py#L58-L60

**Comment by snopoke** [comment]
That is `file_obj._buf.content_length` which can actually be removed now.

File: corehq/blobs/s3db.py
Comment (snopoke) [reply corehq/blobs/s3db.py:88]: That is `file_obj._buf.content_length` which can actually be removed now: 6d2f0c00c9c

**Review by millerdev — COMMENTED** [review]
I fixed some bugs and created a new PR against the original branch in @satyaakam's fork: https://github.com/satyaakam/commcare-hq/pull/1

@snopoke I'm not sure how you were able to push directly to this PR since it's originating from a fork. I'd be interested in learning how you did that.

**Comment by snopoke** [comment]
> I'm not sure how you were able to push directly to this PR since it's originating from a fork. I'd be interested in learning how you did that.

I got given write access to the forked repo

**Review by millerdev — APPROVED** [review]

**Comment by snopoke** [comment]
@millerdev build is passing but will let you do a final review before merging

**Comment by shubham1g5** [comment]
Flagging that this PR incorrectly updated the commcare-translations submodule reverting it back to an older 2.47 version. @orangejenny has made the required correction on master pushing it back to the latest commit.

**Pull request merged**
